### PR TITLE
Add "wrapping" registration flow

### PR DESCRIPTION
**Pull request by jefferai**
This uses a KMS (a wrapper) as a third-party arbitrator of validity. The signed bundle can contain registration information encrypted by a shared KMS. This information consists of the nonce and the public key from the signer of the bundle itself, which means our existing protections against replay (such as the result being encrypted to the signed encryption key ID) apply here as well. If the value is able to be successfully decrypted, the node is authorized.

For multi-hop situations, an interim node can unwrap the value using its wrapper and encrypt the message to the server using its encryption key. Notably, this encrypted value is placed outside the signed bundle. If the server trusts this node it can verify that the node provided the unwrapped values by decrypting using its shared key. Notably, it is still on the server to validate the contained information (nonce and key) after it validates the signature of the bundle.

### CODE REVIEW

File: registration/register_node_led.go [hunk 1]
```diff
@@ -121,7 +122,60 @@ func FetchNodeCredentials(
...
+		err := fmt.Errorf("(%s) error during fetch request validation: %w", op, err)
+		return nil, err
+
+	case len(reqInfo.WrappedRegistrationInfo) > 0 || len(req.RewrappedWrappingRegistrationFlowInfo) > 0:
+		var registrationInfo *types.WrappingRegistrationFlowInfo
+		if len(req.RewrappedWrappingRegistrationFlowInfo) > 0 {
+			encryptingNodeInfo, err := types.LoadNodeInformation(ctx, storage, req.RewrappingKeyId, opt...)
+			if err != nil {
+				err := fmt.Errorf("(%s) error looking up node information for rewrapping key id: %w", op, err)
+				opts.WithLogger.Error(err.Error())
```
Comment (jimlambrt) [inline registration/register_node_led.go:134]: typically not a fan of logging in this sort of module... but this is def needed
in this one.

File: registration/register_node_led.go [hunk 2]
```diff
@@ -222,3 +277,43 @@ func FetchNodeCredentials(
 		ServerEncryptionPublicKeyType:     nodeInfo.ServerEncryptionPrivateKeyType,
 	}, nil
 }
+
+// DecryptWrappedRegistrationInfo is shared functionality for decrypting wrapped
+// registration information that can be used both within registration and during
+// multi-hop contexts
+func DecryptWrappedRegistrationInfo(ctx context.Context, reqInfo *types.FetchNodeCredentialsInfo, opt ...nodeenrollment.Option) (*types.WrappingRegistrationFlowInfo, error) {
+	const op = "nodeenrollment.registration.DecryptWrappedRegistrationInfo"
```
Comment (jimlambrt) [inline registration/register_node_led.go:285]: should you check `reqInfo != nil`?

**Review by jimlambrt — CHANGES_REQUESTED** [review]
Just one question about a nil check (that I'm worried could cause a panic).

File: registration/register_node_led.go (hunk 1 above)
Comment (jefferai) [reply registration/register_node_led.go:134]: Keep in mind that the idea is that it's not generally logging unless you actually pass it a logger. And that could be one that's only turned on at certain times.

However there is basically no other way to understand what's going on when things aren't working. In those situations activating the logging is invaluable.

File: registration/register_node_led.go (hunk 2 above)
Comment (jefferai) [reply registration/register_node_led.go:285]: Yep! Will add.

File: util/toggledlogger/toggledlogger.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+	"github.com/hashicorp/go-hclog"
+)
+
+// ToggledLogger is used to provide an hclog-style interface that can be turned
+// on/off at will. This is useful for providing a logger to this package where
+// you may only want internal authentication logging to happen in specific
+// debugging scenarios.
+//
+// When using With and Named and friends, the returned logger will share the
```
Comment (irenarindos) [inline util/toggledlogger/toggledlogger.go:19]: What's mean by friends?

**Review by irenarindos — APPROVED** [review]

**Pull request merged**